### PR TITLE
Allow v2 Resize to resize longer edge exactly to `max_size`

### PR DESCRIPTION
**Pull request by sidijju**
Addresses new feature request opened in #8358

Looking for feedback regarding v1 compatibility issues / potential workarounds for that.

cc @vfdev-5

### CODE REVIEW

File: test/test_transforms_v2.py [hunk 1]
```diff
@@ -99,7 +99,7 @@ def _script(obj):
         return torch.jit.script(obj)
     except Exception as error:
         name = getattr(obj, "__name__", obj.__class__.__name__)
-        raise AssertionError(f"Trying to `torch.jit.script` '{name}' raised the error above.") from error
+        raise AssertionError(f"Trying to 'torch.jit.script' {name} raised the error above.") from error
```
Comment (NicolasHug) [inline test/test_transforms_v2.py:102]: Let's revert these changes if you don't mind: they are not strictly related to the changes in this PR and thus it makes it a bit harder to do the review as they add noise.
(It's also a matter of preference whether backticks should be used, so it's best to leave that discussion for elsewhere).

Please revert other similar changes as well

File: test/test_transforms_v2.py [hunk 2]
```diff
@@ -810,6 +826,30 @@ def test_max_size_error(self, size, make_input):
...
+            make_image_pil,
+            make_image,
+            make_bounding_boxes,
+            make_segmentation_mask,
+            make_detection_masks,
+            make_video,
+        ],
+    )
+    def test_size_none(self, max_size, device, make_input):
+        check_transform(
```
Comment (NicolasHug) [inline test/test_transforms_v2.py:847]: It seems that this test doesn't add much value over the existing `check_transform`, apart from testing different values of `max_size`? Maybe we can remove it and instead check for the core logic of `size=None, max_size=int` with some hard-coded values of `max_size` and `input_size`, checking the expected shape. Basically, something like this:

```py
    @pytest.mark.parametrize("input_size, max_size, expected_size",
        [
            ((10, 10), 10, (10, 10)),
            ((10, 20), 40, (20, 40)),
            ((20, 10), 40, (40, 20)),
            ((10, 20), 10, (5, 10)),
            ((20, 10), 10, (10, 5)),
        ]
    )
    @pytest.mark.parametrize("make_input", [make_image_tensor, make_image_pil])
    def test_resize_size_none(self, input_size, max_size, expected_size, make_input):
        img = make_input(input_size)
        out = F.resize(img, size=None, max_size=max_size)
        assert F.get_dimensions(out)[-2:] == list(expected_size)

```

File: torchvision/transforms/v2/functional/_geometry.py [hunk 3]
```diff
@@ -159,21 +159,30 @@ def vertical_flip_video(video: torch.Tensor) -> torch.Tensor:
...
 ) -> List[int]:
     if isinstance(size, int):
         size = [size]
+    elif size is None:
+        if isinstance(max_size, int):
+            h, w = canvas_size
+            short, long = (w, h) if w <= h else (h, w)
+            new_short, new_long = int(max_size * short / long), max_size
+            new_w, new_h = (new_long, new_short) if w <= h else (new_short, new_long)
+            size = [new_w, new_h]
```
Comment (NicolasHug) [inline torchvision/transforms/v2/functional/_geometry.py:172]: It seems that this logic should be implemented in `__compute_resized_output_size` instead, because this is where the rest of the size computation logic resides.
Since `__compute_resized_output_size` is also used for `v1` and we don't want v1 to support None, maybe we can just add a flag `allow_size_none=False`  which we set to `True` here in v2 only?

**Review by NicolasHug — COMMENTED** [review]
Thank you @sidijju ! Overall the PR looks great, I just pushed something to avoid overriding the `_extract_params_for_v1_transform` method. I also made a few comments below, LMK what you think.

The macos CI jobs are all red right now, but that's unrelated and can be safely ignored.

File: test/test_transforms_v2.py (hunk 1 above)
Comment (sidijju) [reply test/test_transforms_v2.py:102]: Sure I can do that! Do you have advice on how to avoid getting flake errors from formatting the commit? I added these changes originally because flake would fail due to inclusion of the backticks.

File: test/test_transforms_v2.py [hunk 4]
```diff
@@ -551,7 +551,7 @@ def affine_bounding_boxes(bounding_boxes):
 
 class TestResize:
     INPUT_SIZE = (17, 11)
-    OUTPUT_SIZES = [17, [17], (17,), None, [12, 13], (12, 13)]
+    OUTPUT_SIZES = [17, [17], (17,), [12, 13], (12, 13), None]
```
Comment (sidijju) [inline test/test_transforms_v2.py:554]: I put the value of None in the middle of the list originally because some of the tests (for example `test_kernel_mask` and `test_kernel_video`) use `OUTPUT_SIZES[-1]` as the value for size, and I felt that those tests weren't intended to be run using `size = None`. 

Some of the tests also use `OUTPUT_SIZES[0]` (e.g. test_pil_interpolation_compat_smoke), so I felt that putting None at the beginning of the list would be improper as well.

File: test/test_transforms_v2.py [hunk 5]
```diff
@@ -847,8 +847,7 @@ def test_size_none(self, max_size, device, make_input):
         check_transform(
             transforms.Resize(size=None, max_size=max_size, antialias=True),
             make_input(self.INPUT_SIZE, device=device),
-            # atol=1 due to Resize v2 is using native uint8 interpolate path for bilinear and nearest modes
-            check_v1_compatibility=dict(rtol=0, atol=1),
+            check_v1_compatibility=False,
```
Comment (sidijju) [inline test/test_transforms_v2.py:850]: Should we make this the same as the other `check_v1_compatibility` argument, where we keep the dictionary and only set it to False if size is None?

File: test/test_transforms_v2.py (hunk 2 above)
Comment (sidijju) [reply test/test_transforms_v2.py:847]: Sounds good!

File: torchvision/transforms/v2/functional/_geometry.py (hunk 3 above)
Comment (sidijju) [reply torchvision/transforms/v2/functional/_geometry.py:172]: That makes sense - I didn't want to change the v1 code originally but this solution works. I'll make the change.

File: test/test_transforms_v2.py (hunk 4 above)
Comment (NicolasHug) [reply test/test_transforms_v2.py:554]: Ah good point, I missed that when I did the changes, thanks!

File: test/test_transforms_v2.py (hunk 5 above)
Comment (NicolasHug) [reply test/test_transforms_v2.py:850]: This may not be needed here because `size` is always None in this test, and in any case we'll remove this test and instead use something like suggested in https://github.com/pytorch/vision/pull/8459/files#r1622182968

File: test/test_transforms_v2.py (hunk 1 above)
Comment (NicolasHug) [reply test/test_transforms_v2.py:102]: Hm yeah, I see (a newer version of?) `flake8`  being unhappy locally. It seems like a bug/regression to me. backticks used to be a proper python construct in Python2 and they were dropped in Python 3 in favor of `repr()`, so `flake8` is telling us to use `repr()` instead. But here we're using backticks within a string - there's no reason to error on that, so let's just ignore it.

Comment (sidijju) [reply test/test_transforms_v2.py:102]: Sounds good!

File: test/test_transforms_v2.py (hunk 5 above)
Comment (sidijju) [reply test/test_transforms_v2.py:850]: Sounds good!

**Comment by sidijju** [comment]
Just pushed some changes @NicolasHug! Lmk if you have any thoughts

**Comment by sidijju** [comment]
Just pushed some changes @NicolasHug! Lmk if you have any thoughts
